encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/ArrayToken.java

```diff
@@ -0,0 +1,27 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders;
+
+class ArrayToken<T> extends TypeToken<T> {
+  private TypeToken<?> componentType;
```

Comment:
to make types align a bit better, consider changing to:

```suggestion
class ArrayToken<T> extends TypeToken<T[]> {
  private TypeToken<T> componentType;
```
This might make it harder to implement `TypeToken#of()` but having the types align externally is a useful property to have.

Reply:
Having figured out a way to do `T[]`, but find another easy way to align types a bit better. what do you think? @vkryachko 
```
  public static class ArrayToken<T, C> extends TypeToken<T> {
    private final TypeToken<C> componentType;

    private ArrayToken(TypeToken<C> componentType) {
      this.componentType = componentType;
    }

    @NonNull
    public TypeToken<C> getComponentType() {
      return componentType;
    }
  }
```

Reply:
One thing to consider in the `T,C` approach is that `T` and `C` are completely unconstrained and unrelated, so even though the only sensible instantiation of such an array token would be `ArrayToken<T[], T>`, I can construct arbitrary pairs like `ArrayToken<List<String>, Boolean>`, which arguably does not make much sense.

Whereas with the `T[]` approach it is guaranteed that `ArrayToken<T>`s have components of type `T` while also - at the type level - being `TypeToken<T[]>`s, inherently indicating the fact that they are indeed arrays. wdyt?

Reply:
> Haven't figured out a way to do T[]

We can pair on this one offline if you like

Reply:
Actually, having thought more about it, it might not be such a good idea to have `T[]` as it makes it hard to work with `ArrayToken`s in generic code, i.e. if I have a `TypeToken<T>` there is no way I can correctly cast it to `ArrayToken<???>` as I cannot take away the `array`-ness off of `T`. Maybe for now let's leave it as you have it currently: with a `?` component type

Reply:
Yes, I agree. Let's leave it as `?` type.